README-design.md

```diff
@@ -0,0 +1,55 @@
+This document describes some of the current high level design
+of the project.  It assumes some familiarity with the primary `README.md`.
+
+Builds
+----
+
+Conceptually, each run of `coreos-assembler build` may generate a new "build".
+
+By default, a single build will generate both an OSTree commit and a `-qemu.qcow2`.
+The OSTree commit data is generated via rpm-ostree, using `src/config/manifest.yaml`.
+The QEMU image uses the OSTree commit and the `src/config/image.ks` data, which
+defines the partition layout.
+
+There is a single OSTree repository created by `coreos-assembler init`. The new
+OSTree commit is written into `repo`; if you have configured a ref in the
+manifest, it will be updated.  For more information on OSTree and build systems,
+see [the libostree docs](https://ostree.readthedocs.io/en/latest/manual/buildsystem-and-repos/).
+
+The coreos-assembler concept of a "build" binds together an OSTree commit with
+one or more images that contain the filesystem tree represented by the commit.
+
+Physically, a coreos-assembler build is represented primarily by a new
+subdirectory in `builds/$version`, and inside that directory there's a
+`meta.json` that contains a lot of relevant metadata, including the OSTree
+commit.
```

Comment:
i know in the case of an oscontainer this might not make sense, but in the case of a pure ostree repo model (i.e. fedora coreos for now) it would be kinda cool if we just stored the meta.json content in a detached metadata commit in the ostree repo itself. Then we could store and get all information about a particular commit/version from the repo itself and wouldn't have to manage data separately.

Reply:
That basically implies always creating a new ostree commit, even in the case where just the image changes right?

Which...we could do; just create a new commit that inherits all of the content from the previous one.

But it'd be some new nontrivial logic in the build logic we have today.

Reply:
>That basically implies always creating a new ostree commit, even in the case where just the image changes right?

I'll admit I'm a bit out of my league here on full understanding, but I think the answer to your question depends on how we implement it. I recently had a discussion on twitter where someone mentioned detached metadata (maybe [this is what he was talking about](https://github.com/ostreedev/ostree-releng-scripts/blob/master/ostree-detached-metadata)). I don't really understand exactly what detached metadata is, but it seems like a way to attach information to a commit even after the commit has been created (i.e. not a new commit, but just attaching new information to an existing commit). If that is what it is then we could possibly used detached metadata to add in our build metadata information.

Reply:
Right, GPG signatures use detached metadata today.

But yeah, storing builds in detached metadata would probably make sense if the primary artifact is an ostree repo.